---
title: "How to use the BSgenomeForge package to forge a BSgenome data package"
author:
- name: Atuhurira Kirabo Kakopo
- name: Herve Pages
output:
  html_document:
      toc: true
vignette: |
  %\VignetteIndexEntry{How to use the BSgenomeForge package to forge a BSgenome
   data package}
  %\VignetteEngine{knitr::rmarkdown}
editor_options: 
  markdown: 
    wrap: 80
---

# Introduction

BSgenome data packages are one of the many types of annotation packages
available in [*Bioconductor*](https://bioconductor.org). They contain the
genomic sequences, which comprise chromosome sequences and other DNA sequences,
of a particular genome assembly for a given organism. For example
[BSgenome.Hsapiens.UCSC.hg19](https://bioconductor.org/packages/BSgenome.Hsapiens.UCSC.hg19/)
is a BSgenome data package that contains the genomic sequences of the hg19
genome from [UCSC](https://genome.ucsc.edu/cgi-bin/hgGateway). Users can easily
and efficiently access the sequences, or portions of the sequences, stored in
these packages, via a common API implemented in the
[BSgenome](https://bioconductor.org/packages/BSgenome/) software package.

Bioconductor currently provides more than [100 BSgenome data packages](https://bioconductor.org/packages/release/BiocViews.html#___BSgenome),
for more than 30 organisms. Most of them contain the genomic sequences of
_UCSC genomes_ (i.e. genomes supported by the UCSC Genome Browser) or
_NCBI assemblies_. The packages are used in various Bioconductor workflows, as
well as in man page examples and vignettes of other Bioconductor packages,
typically in conjunction with tools available in the
[BSgenome](https://github.com/Bioconductor/BSgenome) and
[Biostrings](https://bioconductor.org/packages/Biostrings) software packages.
New BSgenome data packages get added on a regular basis, based on user demand.

The current process of forging a BSgenome data package requires one to download
the sequence data files, either as a single twoBit file (e.g. musFur1.2bit) or
in a collection of compressed FASTA files. The user then needs to prepare the
BSgenome data package seed file, and then call `forgeBSgenomeDataPkg` function
on the seed file to forge the package. This is quite a convoluted task that can
be an inconvenience for the average user.

The BSgenomeForge package simplifies the whole process, by allowing the user to
use one function to create a BSgenome data package. It is comprised of two major
functions: `forgeBSgenomeDataPkgfromNCBI` and `forgeBSgenomeDataPkgfromUCSC`
which allow the user to forge BSgenome data packages with sequence data files
sourced from NCBI or UCSC respectively.

# Basic usage

## Using `forgeBSgenomeDataPkgFromNCBI()`

1.Create a BSgenome data package for a NCBI assembly registered in GenomeInfoDb.
This is the organism *Felis catus*.

```{r, eval = FALSE}
forgeBSgenomeDataPkgFromNCBI(assembly_accession="GCF_000181335.3",
                             organism="Felis catus",
                             pkg_maintainer="Jane Doe <janedoe@gmail.com>")
```

2.Create a BSgenome data package for a NCBI assembly NOT registered in
GenomeInfoDb. This is the organism *Acidianus infernus*.

```{r, eval = FALSE}
forgeBSgenomeDataPkgFromNCBI(assembly_accession="GCA_009729545.1",
                             organism="Acidianus infernus",
                             pkg_maintainer="Jane Doe <janedoe@gmail.com>")
```

## Using `forgeBSgenomeDataPkgFromUCSC()`

# Final steps

`forgeBSgenomeDataPkgfromNCBI` or `forgeBSgenomeDataPkgfromUCSC` returns the
path to the created package at the end of its execution. This can be used to
find the package location, and afterwards build the package source tarball with

R CMD build \<pkgdir\>

where \<pkgdir\> is the path to the source tree of the package. Then check the
package with

R CMD check \<tarball\>

where \<tarball\> is the path to the tarball produced by R CMD build. Finally
install the package with

R CMD INSTALL \<tarball\>

and it is ready for use!

# sessionInfo()

```{r, eval = TRUE}
sessionInfo()
```
